xmtp_mls/src/groups/message_history.rs

```diff
@@ -178,37 +269,162 @@ where
+    pub async fn get_pending_history_request(
+        &self,
+    ) -> Result<Option<(String, String)>, MessageHistoryError> {
+        let sync_group = self.get_sync_group()?;
+
+        // sync the group
+        sync_group.sync(self).await?;
 
-        let requests = conn.get_group_messages(
-            sync_group_id,
+        let messages = sync_group.find_messages(
+            Some(GroupMessageKind::Application),
+            None,
+            None,
             None,
             None,
+        )?;
+        let last_message = messages.last();
+
+        let history_request: Option<(String, String)> = if let Some(msg) = last_message {
+            let message_history_content =
+                serde_json::from_slice::<MessageHistoryContent>(&msg.decrypted_message_bytes)?;
+            match message_history_content {
+                // if the last message is a request, return its request ID and pin code
+                MessageHistoryContent::Request(request) => {
+                    Some((request.request_id, request.pin_code))
+                }
+                _ => None,
+            }
+        } else {
+            None
+        };
+
+        Ok(history_request)
+    }
+
+    pub async fn reply_to_history_request(
+        &self,
+    ) -> Result<MessageHistoryReply, MessageHistoryError> {
+        let pending_request = self.get_pending_history_request().await?;
+
+        if let Some((request_id, _)) = pending_request {
+            let reply = self.prepare_history_reply(&request_id).await?;
+            self.send_history_reply(reply.clone().into()).await?;
+            return Ok(reply.into());
+        }
+
+        Err(MessageHistoryError::NoPendingRequest)
+    }
+
+    pub async fn get_latest_history_reply(
+        &self,
+    ) -> Result<Option<MessageHistoryReply>, MessageHistoryError> {
+        let sync_group = self.get_sync_group()?;
+
+        // sync the group
+        sync_group.sync(self).await?;
+
+        let messages = sync_group.find_messages(
+            Some(GroupMessageKind::Application),
+            None,
+            None,
+            None,
+            None,
+        )?;
+
+        let last_message = messages.last();
+
+        let reply: Option<MessageHistoryReply> = match last_message {
+            Some(msg) => {
+                // if the message was sent by this installation, ignore it
+                if msg
+                    .sender_installation_id
+                    .eq(&self.installation_public_key())
+                {
+                    None
+                } else {
+                    let message_history_content = serde_json::from_slice::<MessageHistoryContent>(
+                        &msg.decrypted_message_bytes,
+                    )?;
+                    match message_history_content {
+                        // if the last message is a reply, return it
+                        MessageHistoryContent::Reply(reply) => Some(reply),
+                        _ => None,
```

Comment:
Should log the error from here

Reply:
won't the error surface with the `?` on line 359? i re-did the `serde_json` errors in a recent commit.